cdm/core/src/main/java/ucar/nc2/iosp/hdf5/H5header.java

```diff
@@ -1739,18 +1739,83 @@ private boolean makeVariableShapeAndType(Variable v, MessageDatatype mdt, Messag
+      // look for a case 1 enum type that is structurally the same as the variable's enum type.
+      // If such exists, then use that.
+      // Step 3: Otherwise, create a new enum type and use that. The new enum type
+      // will have these properties:
+      // a. It is defined in the same group as the variable
+      // b. It has a mutated name similar to the variable's name, namely <variablename>_enum_t.
+
+      EnumTypedef actualEnumTypedef = null; // The final chosen EnumTypedef
       Group ncGroup = v.getParentGroupOrRoot();
-      EnumTypedef local = new EnumTypedef(mdt.enumTypeName, mdt.map);
-      EnumTypedef enumTypedef =
-          ncGroup.getEnumTypedefs().stream().filter((e) -> e.equalsMapOnly(local)).findFirst().orElse(local);
-      if (enumTypedef != null) {
-        // if found, make sure it is added to the group
-        ncGroup.addEnumeration(enumTypedef);
-      } else { // if shared object, wont have a name, shared version gets added later
-        enumTypedef = new EnumTypedef(mdt.enumTypeName, mdt.map);
-        ncGroup.addEnumeration(enumTypedef);
-      }
-      v.setEnumTypedef(enumTypedef);
+
+      // Step 1:
+      // See if an independent enum type already exists with the same name
+      EnumTypedef candidate = ncGroup.findEnumeration(mdt.enumTypeName, true);
+      if (candidate != null) {
+        // There is an independent type, so use it.
+        actualEnumTypedef = candidate;
+      }
+
+      // Step 2:
+      // See if an independent enum type already exists that is structurally similar.
+      if (actualEnumTypedef == null && mdt.enumTypeName.equals(v.getShortName())) {
+        // Materialize a enum type def for search purposes; name is irrelevant
+        EnumTypedef template = new EnumTypedef(mdt.enumTypeName, mdt.map);
+        // Search for a structurally similar enum type def
+        candidate = ncGroup.findSimilarEnumTypedef(template, true);
+        if (candidate != null) {
+          // There is an independent type, so use it.
+          actualEnumTypedef = candidate;
+        }
+      }
+
+      // Step 3: Create an independent type
+      if (actualEnumTypedef == null) {
+        String newname = null;
+        if (mdt.enumTypeName.equals(v.getShortName())) {
+          // Create mutated name to avoid name conflict
+          newname = mdt.enumTypeName + "_enum_t";
+        } else {
+          newname = mdt.enumTypeName;
+        }
+        actualEnumTypedef = new EnumTypedef(newname, mdt.map);
+        // Add to the current group(builder)
+        ncGroup.addEnumeration(actualEnumTypedef);
+      }
+
+      if (actualEnumTypedef == null) {
+        log.warn("Missing EnumTypedef: {}", mdt.enumTypeName);
+        throw new IllegalStateException("Missing EnumTypedef: " + mdt.enumTypeName);
+      }
+
+      // associate with the variable
+      v.setEnumTypedef(actualEnumTypedef);
+      /*
+       * Group ncGroup = v.getParentGroupOrRoot();
```

Comment:
maybe the commented out block can go now

Reply:
Done